crates/warg/src/package/validate.rs

```diff
@@ -156,52 +163,70 @@ impl Validator {
         Self::default()
     }
 
+    /// Gets the current root of the validator.
+    ///
+    /// Returns `None` if no records have been validated yet.
+    pub fn root(&self) -> &Option<Root> {
+        &self.root
+    }
+
     /// Validates an individual package record.
     ///
     /// It is expected that `validate` is called in order of the
     /// records in the log.
+    ///
+    /// This operation is transactional: if any entry in the record
+    /// fails to validate, the validator state will remain unchanged.
     pub fn validate(
         &mut self,
         envelope: &ProtoEnvelope<model::PackageRecord>,
     ) -> Result<Vec<DynHash>, ValidationError> {
-        let record = envelope.as_ref();
+        let snapshot = self.snapshot();
 
-        // Validate previous hash
-        self.validate_record_hash(record)?;
+        let mut op = || {
```

Comment:
Perhaps a `transact` method that takes an `FnMut` and wraps a call to it with the rollback behavior to make it more explicit, then we can either just pass in this closure or extract it out to another method?

Reply:
I don't think we need a generic transact function since I don't know when else we'd need to transact.
I just think that this try-and-rollback logic should be separate from the core validate logic for clarity.

e.g.
```rust
fn validate(...) -> ... {
   let snapshot = self.snapshot();
   let result = self.validate_inner(...);
   if result.is_err() {
      self.rollback(snapshot);
   }
   result
}

fn validate_inner(...) -> ... {
   <current validate logic here>
}
```